### ISSUE TITLE
Dictionary<int,*> Serialization/Deserialization issue

### ISSUE THREAD

**Issue by mircoboschi** [obarlik/Json.Net#8]
Serialization of Dictionaries with `int` key leads to json key treated as string.
This causes deserialization problem, because it tries to deserialize a string into int (key).
Here is a small example to reproduce the issue:
```
Dictionary<int, string> dict_test = new Dictionary<int, string>();
dict_test.Add(1, "hello");
string serialized_json = Json.Net.JsonNet.Serialize(dict_test);
Console.WriteLine(serialized_json);//Output: {"1","hello"}
Dictionary<int,string> deserialized_dict = Json.Net.JsonNet.Deserialize<Dictionary<int, string>>(serialized_json);//throws exception
```
Exploring the source code, seems that the key is designed to be only a string
```
               else if (objectType.IsGenericType
                      && objectType.GenericTypeArguments.Length == 2
                      && objectType == (genericType = typeof(KeyValuePair<,>).MakeGenericType(objectType.GenericTypeArguments)))
                {
                    var k = genericType.GetProperty("Key").GetValue(obj).ToString();
                    var v = genericType.GetProperty("Value").GetValue(obj);

                    Write(strConverter(k));//This cause the int->string conversion, and generates the  "1" portion of json output
                    Write(":");
                    Serialize(v, options);
                    return;
```
If we deserialize a hand-fixed string like `{1,"hello"}` we get an exception because it tries to cast the int key to a "name" string.
```
                while (NextChar!='}')
                {
                    var name = (string)FromJson(nameType);

                    SkipWhite();
                    Match(":");
```

**Comment by obarlik** [comment]
Thanks for reporting this issue, I'm on it.

**Comment by obarlik** [comment]
Fixed this issue in version 1.23
